### PR TITLE
Tf/transformer block

### LINKED ISSUE #719: [TASK] Add Huggingface transformers as a backend
### Option 1
`pip install merlin-models[tensorflow-hf]`

### Option 2
For TF:
`pip install merlin-models[tensorflow, transformers]`
For PyTorch:
`pip install merlin-models[pytorch, transformers]`

### LINKED ISSUE #721: Port TransformerBlock

### PR DESCRIPTION
**Pull request by sararb**
Fixes #721 

### Goals :soccer:
- [x] Port TransformerBlock from transformers4rec
- [x] Implement specific HF transformer blocks: Bert, Albert, Roberta, XLNet, and GPT2

### Implementation Details :construction:
- Port the TransformerBlock from T4Rec and refactor it based on the latest HuggingFace code
- Add `pre` layer to support preparing additional inputs of the transformer model such as `attention_mask` or `head_mask`. This layer has access to the sequence of input embeddings and raw features. 
- Add a post with an `output_fn` function to process the [TFBaseModelOutputWithPoolingAndCrossAttentions](https://github.com/huggingface/transformers/blob/0b567aa430e9b1b56254de4775db5868104d34e5/src/transformers/modeling_tf_outputs.py#L125) return by HF layer and return the desired tensor(s). 
- Add custom transformers blocks (bert, albert..) with the `build_transformer_layer` method that prepares the HF configuration class with unified names of the required parameters `(d_model, n_head, n_layer,max_seq_length)`

### Testing Details :mag:
Add unit tests of the transformer block: 

- Test the TransformerBlock as a sequence encoder
- Test pooling the sequence of hidden states returned by the transformer model to a 2-D tensor
- Test different transformer encoding models: Bert, Albert, XLnet, RoBerta, GPT2.
- Test a TransformerBlock in a classification model
- Test save and load back classification model with transformer encoder block
- Test prepare additional inputs to the TransformerBlock

### Open questions/remarks 

- HuggingFace does not support ragged representation, so the input batch of interaction embeddings is converted to a dense tensor within the TransformerBlock call. 
- In case of masking is being used, How to access the boolean mask tensor?

### CODE REVIEW

**Comment by sararb** [comment]
A quick comment about the failing test related to the RoBerta model:  An out-of-index error raised by the position-ids  tensor defined ([here](https://github.com/huggingface/transformers/blob/ca485e562b675341409e3e27724072fb11e10af7/src/transformers/models/roberta/modeling_tf_roberta.py#L163)). As the padding_idx (0 in our case) is overridden to `padding_idx=1` in the [RobertaEmbeddings](https://github.com/huggingface/transformers/blob/ca485e562b675341409e3e27724072fb11e10af7/src/transformers/models/roberta/modeling_tf_roberta.py#L84) class. A quick workaround would be to increase the `max_position_embeddings`, but I don't know why the padding_idx parameter of RoBerta is fixed to 1 in HF.

**Comment by sararb** [comment]
Closing this PR and opening an updated one including the latest main changes
